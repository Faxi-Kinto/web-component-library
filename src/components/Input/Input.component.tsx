import React, { useState, ChangeEvent, useMemo } from 'react';
import * as Styled from './Input.styles';

/**
 * @name Input
 *
 * Straightforward input field
 *
 * @returns {JSX}
 */

export type InputProps = {
  name?: string;
  value?: string;
  placeholder?: string;
  label?: string;
  onChange?: (value: string) => void;
  error?: boolean;
  type?: string;
  onHandleCapsLock?: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  autoFocus?: boolean;
};

const Input: React.FC<InputProps> = (props: InputProps): JSX.Element => {
  const {
    name,
    value,
    placeholder,
    label,
    onChange,
    error = false,
    type = 'text',
    onHandleCapsLock,
    autoFocus,
    ...rest
  } = props;
  const [inputValue, setInputValue] = useState('');

  const currentValue = onHandleCapsLock ? value?.replace(/\s/g, '') : value;

  const htmlFor = useMemo(() => {
    if (name)
      return {
        htmlFor: name,
      };
  }, [name]);

  const id = useMemo(() => {
    if (name)
      return {
        id: name,
        name,
      };
  }, [name]);

  const checkCapsLockStatus = (
    event: React.KeyboardEvent<HTMLInputElement>
  ) => {
    if (onHandleCapsLock) {
      onHandleCapsLock(event);
    }
  };

  function handleOnChange(event: ChangeEvent<HTMLInputElement>): void {
    if (onChange) {
      onChange(event.target.value);
    } else if (onHandleCapsLock) {
      setInputValue(event.target.value.replace(/\s/g, ''));
    } else {
      setInputValue(event.target.value);
    }
  }

  return (
    <Styled.InputContainer>
      {label && <label {...htmlFor}>{label}</label>}
      <input
        {...id}
        {...rest}
        className={error ? 'input--error' : ''}
        type={type}
        value={currentValue || inputValue}
        placeholder={placeholder}
        autoFocus={autoFocus}
        onChange={handleOnChange}
        onKeyPress={(event: React.KeyboardEvent<HTMLInputElement>) =>
          checkCapsLockStatus(event)
        }
      />
    </Styled.InputContainer>
  );
};

export default Input;
