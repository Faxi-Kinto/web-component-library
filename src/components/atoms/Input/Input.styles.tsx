import styled from 'styled-components';
import { marginChildren, pxToRem, padding } from '@faxi/web-css-utilities';

export const InputContainer = styled.div`
  width: 100%;
  ${marginChildren(`0 0 ${pxToRem('10px')} 0`)};

  > input {
    width: 100%;
    height: ${pxToRem('45px')};
    ${padding(pxToRem('8px'), pxToRem('16px'))};
    outline: none;

    &:focus,
    &.input--error {
      box-shadow: none;
    }
  }
`;
