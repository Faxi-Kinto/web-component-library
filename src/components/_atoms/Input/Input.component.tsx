import React, { useState, ChangeEvent, useMemo } from 'react';
import * as Styled from './Input.styles';
import Label from '../Label';
import classNames from 'classnames';

/**
 * @name Input
 *
 * @returns {JSX}
 */

export type InputProps = {
  autoFocus?: boolean;
  autoComplete?: 'on' | 'off';
  className?: string;
  disabled?: boolean;
  error?: boolean;
  id?: string;
  label?: string;
  name?: string;
  placeholder?: string;
  type?: string;
  value?: string;
  onChange?: (value: string) => void;
  onClick?: () => void;
  onFocus?: () => void;
  onBlur?: () => void;
};

const Input: React.FC<InputProps> = (props: InputProps) => {
  const {
    autoFocus,
    autoComplete,
    className,
    disabled = false,
    error = false,
    id,
    label,
    name,
    placeholder,
    type = 'text',
    value,
    onChange,
    onClick,
    onFocus,
    onBlur,
    ...rest
  } = props;
  const [inputValue, setInputValue] = useState<string>('');

  const htmlFor = useMemo(() => {
    if (id)
      return {
        htmlFor: id,
      };
    else if (name)
      return {
        htmlFor: name,
      };
    return {};
  }, [id, name]);

  const ID = useMemo(() => {
    if (id)
      return {
        id: id,
      };
    else if (name)
      return {
        id: name,
      };
    return {};
  }, [id, name]);

  const handleOnChange = (event: ChangeEvent<HTMLInputElement>): void => {
    if (onChange) {
      onChange(event.target.value);
    } else {
      setInputValue(event.target.value);
    }
  };

  return (
    <Styled.InputContainer>
      {label && <Label {...htmlFor}>{label}</Label>}
      <input
        autoFocus={autoFocus}
        autoComplete={autoComplete}
        className={classNames([
          'input',
          { 'input--error': error },
          { 'input--disabled': disabled },
          className,
        ])}
        disabled={disabled}
        {...ID}
        name={name}
        placeholder={placeholder}
        type={type}
        value={value || inputValue}
        onClick={onClick}
        onChange={handleOnChange}
        onFocus={onFocus}
        onBlur={onBlur}
        {...rest}
      />
    </Styled.InputContainer>
  );
};

export default Input;
