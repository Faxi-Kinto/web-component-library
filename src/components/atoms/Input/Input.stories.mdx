import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import { withKnobs, boolean, text } from '@storybook/addon-knobs';
import Input from './';

<Meta title="Core|Atoms/Input" />

# Input Component

Straithforward input component. It can be simple, or with placeholder.
When an error occurs its border changes color

## Implementation

#### Default

<Preview>
  <Story name="default" decorators={[withKnobs]}>
    <Input
      error={boolean('Error', false)}
      placeholder={text('Placeholder text', '')}
    />
  </Story>
</Preview>

#### With Error

<Preview>
  <Story name="With error">
    <Input error={true} />
  </Story>
</Preview>

#### With Placeholder

<Preview>
  <Story name="With a placeholder">
    <Input placeholder="Type value here" error={false} />
  </Story>
</Preview>

#### With Label

<Preview>
  <Story name="With a label">
    <Input label="With label" name="labeled input" error={false} />
  </Story>
</Preview>

#### With Both Placeholder & Label

<Preview>
  <Story name="With placeholder & label">
    <Input
      label="With label"
      placeholder="and also placeholder"
      name="test input"
      error={false}
    />
  </Story>
</Preview>

## Input Props

<Props of={Input} />

## Notes

- Default error value is set to false, so if omitted the field will render without error border

- With onChange prop you can set the state to the inputField from parent components.

- onnHandleCapslook prop is used for the password molecule, to inform component when caps-lock is pressed.
