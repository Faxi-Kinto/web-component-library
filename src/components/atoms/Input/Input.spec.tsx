import React from 'react';
import renderer from 'react-test-renderer';
import Input from './';

test('Input renders correctly', () => {
  const component = renderer.create(
    <Input value="Test value" onChange={(): null => null} error />
  );

  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
