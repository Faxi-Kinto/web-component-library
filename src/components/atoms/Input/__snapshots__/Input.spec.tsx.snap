// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders correctly 1`] = `
.c0 {
  width: 100%;
}

.c0 > * {
  margin: 0 0 0.625rem 0;
}

.c0 > :last-child {
  margin: 0;
}

.c0 > input {
  width: 100%;
  height: 2.8125rem;
  padding: 0.5rem 1rem 0.5rem 1rem;
  outline: none;
}

.c0 > input:focus,
.c0 > input.input--error {
  box-shadow: none;
}

<div
  className="c0"
>
  <input
    className="input--error"
    onChange={[Function]}
    onKeyPress={[Function]}
    type="text"
    value="Test value"
  />
</div>
`;
